stream.go

```diff
@@ -0,0 +1,120 @@
+package dht
+
+import (
+	"context"
+	"sync"
+
+	pbio "github.com/gogo/protobuf/io"
+	pb "github.com/libp2p/go-libp2p-kad-dht/pb"
+	inet "github.com/libp2p/go-libp2p-net"
+	peer "github.com/libp2p/go-libp2p-peer"
+	"golang.org/x/xerrors"
+)
+
+func (dht *IpfsDHT) newStream(ctx context.Context, p peer.ID) (*stream, error) {
+	s, err := dht.host.NewStream(ctx, p, dht.protocols...)
+	if err != nil {
+		return nil, xerrors.Errorf("opening stream: %w", err)
+	}
+	ps := &stream{
+		stream: s,
+		w:      newBufferedDelimitedWriter(s),
+		r:      pbio.NewDelimitedReader(s, inet.MessageSizeMax),
+		m:      make(chan chan *pb.Message, 1),
+	}
+	go func() {
+		ps.reader()
+		dht.streamPool.delete(ps, p)
+		ps.reset()
+	}()
+	return ps, nil
+}
+
+type stream struct {
+	stream interface {
+		Reset() error
+	}
+	w bufferedWriteCloser
+	r pbio.ReadCloser
+
+	// Synchronizes m and readerErr.
+	mu sync.Mutex
+	// Receives channels to send responses on.
+	m         chan chan *pb.Message
+	readerErr error
+}
+
+func (me *stream) reset() {
+	me.stream.Reset()
+}
+
+func (me *stream) send(m *pb.Message) (err error) {
```

Comment:
We always `send` replies, right? Looking at this file only, one thinks the `send`/`request` duality is for fire-and-forget and RPC, respectively.

However, that makes no sense because the readLoop always expects a queued reply channel, which implies that we only ever use this in RPC fashion. Hence I believe `send` is only used in the context of returning a reply, and this will be less misleading:

```suggestion
func (me *stream) sendReply(m *pb.Message) (err error) {
```

Reply:
No, `send` is used for send-only messages, and the request part of `IPFSDht.sendRequest`. We do not use `send` to send responses, that's an entirely different part of the code base I think.